### PR TITLE
New documentable model

### PR DESCRIPTION
**Pull request by kamilok1965**
Do not merge

### CODE REVIEW

File: core/src/main/kotlin/model/Documentable.kt [hunk 1]
```diff
@@ -5,231 +5,273 @@ import org.jetbrains.dokka.model.doc.DocumentationNode
...
-    classlikes = classlikes,
-    expected = expected,
-    actual = actual,
-    extra = extra
-), WithVisibility
+    override val expect: Class?,
+    override val actual: PlatformDependent<Class>,
+    override val visibility: PlatformDependent<Visibility>,
+    override val companion: List<Object>,
+    override val generics: PlatformDependent<Classlike>,
```
Comment (Kordyjan) [inline core/src/main/kotlin/model/Documentable.kt]: I don't think `ClassLike` gives us information we need here. Maybe `TypeWrapper`?

File: core/src/main/kotlin/model/Documentable.kt [hunk 2]
```diff
@@ -5,231 +5,273 @@ import org.jetbrains.dokka.model.doc.DocumentationNode
...
+    abstract val children: List<Documentable>
+    abstract val documentation: PlatformDependent<DocumentationNode>
+    abstract val original: PlatformDependent<Documentable>
+
+    override fun toString(): String =
+        "${javaClass.simpleName}($dri)" //+ briefDocTagString.takeIf { it.isNotBlank() }?.let { " [$it]" }.orEmpty()
+
+    override fun equals(other: Any?) = other is Documentable && this.dri == other.dri
+
+    override fun hashCode() = dri.hashCode()
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply core/src/main/kotlin/model/Documentable.kt]: Added a link to this comment for now

File: core/src/main/kotlin/model/Documentable.kt (hunk 1 above)
Comment (kamilok1965) [reply core/src/main/kotlin/model/Documentable.kt]: Changed to `TypeWrapper` for now and added a TODO to fix this, preferably with PR https://github.com/Kotlin/dokka/issues/659

File: core/src/main/kotlin/model/Documentable.kt [hunk 3]
```diff
@@ -5,231 +5,268 @@ import org.jetbrains.dokka.model.doc.DocumentationNode
...
+
+interface WithGenerics {
+    val generics: PlatformDependent<TypeWrapper> // TODO: fix the generics
+}
+
+interface Callable : WithVisibility, WithType, WithAbstraction, WithExpectActual {
+    val receiver: PlatformDependent<Parameter>
+}
+
+interface CanBeSupertype
```
Comment (sellophane) [inline core/src/main/kotlin/model/Documentable.kt]: maybe just "isOpen" ?

**Pull request merged**